File: js/day-4/s.js
const start = Date.now()

import { readFileReturnLines } from '../utils/read-file-return-lines.mjs'

class Graph {
  constructor (lines) {
    this.g = []
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]
      this.g[i] = line.split('')
    }
  }

  get (x, y) {
    if (x < 0 || x >= this.g.length || y < 0 || y >= this.g[0].length) {
      return null
    }
    return this.g[x][y]
  }

  n (x, y) {
    return this.get(x - 1, y)
  }

  s (x, y) {
    return this.get(x + 1, y)
  }

  e (x, y) {
    return this.get(x, y + 1)
  }

  w (x, y) {
    return this.get(x, y - 1)
  }

  ne (x, y) {
    return this.get(x - 1, y + 1)
  }

  nw (x, y) {
    return this.get(x - 1, y - 1)
  }

  se (x, y) {
    return this.get(x + 1, y + 1)
  }

  sw (x, y) {
    return this.get(x + 1, y - 1)
  }

  getAllAdjacents () {
    const adjacents = []
    for (let i = 0; i < this.g.length; i++) {
      for (let j = 0; j < this.g[i].length; j++) {
        adjacents.push([this.g[i][j], this.getAdjacentsArr(i, j)])
      }
    }
    return adjacents
  }

  getAdjacentsArr (x, y) {
    return [
      this.n(x, y),
      this.s(x, y),
      this.e(x, y),
      this.w(x, y),
      this.ne(x, y),
      this.nw(x, y),
      this.se(x, y),
      this.sw(x, y)
    ]
  }

  getAdjacents (x, y) {
    return {
      n: this.n(x, y),
      s: this.s(x, y),
      e: this.e(x, y),
      w: this.w(x, y),
      ne: this.ne(x, y),
      nw: this.nw(x, y),
      se: this.se(x, y),
      sw: this.sw(x, y),
      arr: [
        this.n(x, y),
        this.s(x, y),
        this.e(x, y),
        this.w(x, y),
        this.ne(x, y),
        this.nw(x, y),
        this.se(x, y),
        this.sw(x, y)
      ]
    }
  }

  print () {
    for (let i = 0; i < this.g.length; i++) {
      console.log(this.g[i].join(''))
    }
  }

  remove (lift = false) {
    let removed = 0
    for (let i = 0; i < this.g.length; i++) {
      for (let j = 0; j < this.g[i].length; j++) {
        if (this.g[i][j] !== '@') continue
        const adjArr = this.getAdjacentsArr(i, j)
        if (adjArr.filter(val => val == '@').length < 4) {
          removed++
          if (lift) this.g[i][j] = '.'
        }
      }
    }
    return removed
  }
}

const lines = readFileReturnLines('../../inputs/day-4/input.txt')

const graph = new Graph(lines)

const ans1 = graph.remove()

let cur = 0
let ans2 = 0

do {
  cur = graph.remove(true)
  ans2 += cur
} while (cur > 0)

console.log(`Answer1: ${ans1}`)
console.log(`Answer2: ${ans2}`)
console.log('Execution time:', Date.now() - start, 'ms')

File: py/day-5/s.py
import os
import time
start_time = time.time()



def read(filename="../../inputs/day-5/input.txt"):
    current_dir = os.path.dirname(os.path.abspath(__file__))
    file_path = os.path.join(current_dir, filename)
    with open(file_path, "r") as f:
        return f.read()
    

lines = read().splitlines()
ranges = []
ids_to_check = []
for line in lines:
  if line.strip() == "":
    current_index = lines.index(line)
    ids_to_check.extend(lines[current_index + 1:])
    break
  start, end = map(int, line.split("-"))
  ranges.append((start, end))


ans2 = 0
i = 0
merged_range_indexes = []
while i < len(ranges):
  range_to_check = ranges[i]
  j = i + 1
  while j < len(ranges):
    next_range = ranges[j]
    if range_to_check[1] >= next_range[0] and range_to_check[1] <= next_range[1]:
      next_range_list = list(next_range)
      next_range_list[0] = min(range_to_check[0], next_range[0])
      ranges[j] = tuple(next_range_list)
      merged_range_indexes.append(i)
    elif range_to_check[0] >= next_range[0] and range_to_check[0] <= next_range[1]:
      next_range_list = list(next_range)
      next_range_list[1] = max(range_to_check[1], next_range[1])
      ranges[j] = tuple(next_range_list)
      merged_range_indexes.append(i)
    elif range_to_check[0] <= next_range[0] and range_to_check[1] >= next_range[1]:
      next_range_list = list(next_range)
      next_range_list[0] = range_to_check[0]
      next_range_list[1] = range_to_check[1]
      ranges[j] = tuple(next_range_list)
      merged_range_indexes.append(i)
    j+=1
  i += 1


# Filter out ranges whose indexes are in merged_range_indexes
filtered_ranges = [r for idx, r in enumerate(ranges) if idx not in merged_range_indexes]

for r in filtered_ranges:
  ans2 += r[1] - r[0] + 1

ans1 = 0
for id in ids_to_check:
  id_num = int(id)
  found = False
  for r in filtered_ranges:
    if r[0] <= id_num <= r[1]:
      found = True
      break
  if found:
    ans1 += 1



print(f"Ans1: {ans1}")
print(f"Ans2: {ans2}")
print(f"Execution time: {(time.time() - start_time) * 1000:.4f} ms")
